### PR TITLE
Vltt 1

### PR DESCRIPTION
**Pull request by vltt**

### CODE REVIEW

File: vltt/assignment1/Q1.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                cur = "";
+            }
+            if (i == s.length()) return;
+        } 
+        else if (is_case_insensitive) cur += tolower(s[i]);
+        else cur += s[i];
+    }
+}
+
+bool is_anagram(std::string &a, std::string &b, bool ch) {
```
[the remark this answers is not in the document]
Comment (kulizhsy) [reply vltt/assignment1/Q1.cpp]: What if the strings are `"Hello." `and `"Hello!"`, your code would not consider them to be anagrams. Should it?

File: vltt/assignment1/Q1.cpp [hunk 2]
```diff
@@ -6,52 +6,53 @@
...
+//is_case_sensitive == 1 -  sensitive, is_case_sensitive == 0 - insensitive
 
-void split_by_word(std::string &s, std::unordered_map<std::string, int> &m, bool is_case_insensitive) {
+
+void split_by_word(std::string &sentence, std::unordered_map<std::string, int> &word_counts, bool is_case_sensitive) {
     std::string cur;
-    for (int i = 0; i <= s.length(); i++) {
-        if (i == s.length() || isspace(s[i])) {
+    for (int i = 0; i <= sentence.length(); i++) {
+        if (i == sentence.length() || !isalpha(sentence[i]) && !isdigit(sentence[i])) {
```
Comment (vltt) [inline vltt/assignment1/Q1.cpp]: Now "hello!" and "hello." is anagram

File: vltt/assignment1/Q1.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#include <vector>
+#include <iomanip>
+#include <string>
+#include <unordered_set>
+#include <unordered_map>
+
+//is_case_sensitive == 1 -  sensitive, is_case_sensitive == 0 - insensitive
+
+
+void split_by_word(std::string &sentence, std::unordered_map<std::string, int> &word_counts, bool is_case_sensitive) {
```
Comment (kulizhsy) [inline vltt/assignment1/Q1.cpp]: You should always pass non-trivial (anything other than a single numeric value or a boolean) objects as `const &`. The benefit of `const` is that it does not allow you to modify the passed-in object, creating a nice contract between the caller and the function that whatever is passed in will be left untouched. 

It's actually expected that whatever is not marked as `const` in the function signature WILL be modified and you would typically either copy the original value before calling such a function or only use throw-away values you don't care about.

So this line should be:

`void split_by_word(const std::string &sentence, std::unordered_map<std::string, int> &word_counts, bool is_case_sensitive) {`

File: vltt/assignment1/Q1.cpp [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                cur = "";
+            }
+            if (i == sentence.length()) return;
+        }
+        else if (is_case_sensitive) cur += sentence[i];
+        else cur += tolower(sentence[i]);
+    }
+}
+
+bool is_anagram(std::string &sentence_a, std::string &sentence_b, bool is_case_sensitive) {
```
Comment (kulizhsy) [inline vltt/assignment1/Q1.cpp]: Should be:

`bool is_anagram(const std::string &sentence_a, const std::string &sentence_b, bool is_case_sensitive)`

File: vltt/assignment1/Q2.cpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#include <iostream>
+
+struct node {
+    node *next = nullptr;
+    int value;
+};
+
+node *kth_elem(node *root, int k) {
```
Comment (kulizhsy) [inline vltt/assignment1/Q2.cpp]: Should be:

`node *kth_elem(const node &root, int k)`

**Pull request merged**